File: datagrok-celery-task/datagrok_celery_task/func_call.py
from enum import Enum
from typing import Union, Any, Dict, List


class Type(str, Enum):
    INT = 'int'
    STRING = 'string'
    BOOL = 'bool'
    FLOAT = 'double'
    BIG_INT = 'bigint'
    DATA_FRAME = 'dataframe'
    DATE_TIME = 'datetime'
    FILE = 'file'
    BLOB = 'blob'


class FuncCallStatus(str, Enum):
    RUNNING = 'Running'
    QUEUED = 'Queued'
    COMPLETED = 'Completed'
    ERROR = 'Error'
    CANCELED = 'Canceled'
    CANCELING = 'Canceling'
   

class FuncCallParam:
    def __init__(self, **kwargs):
        self.name = kwargs.get("name")
        self.property_type = Type(kwargs.get("property_type"))
        self.value = kwargs.get("value")
        self.is_input = kwargs.get("is_input")

    @property
    def is_streamable(self) -> bool:
        return self.property_type == Type.DATA_FRAME or self.property_type == Type.BLOB or self.property_type == Type.FILE


class FuncCall:
    default_binary_batch_size: int = 2048000

    def __init__(self, call: Dict[str, Any]):
        self.id = call.get("id")
        self.func: Dict[str, Any] = call.get("func", {})
        self.options: Dict[str, Any] = call.get("options", {})
        self.aux: Dict[str, Any] = call.get("aux", {})
        self.status = FuncCallStatus.RUNNING

        parameter_values: Dict[str, Any] = call.get("parameterValues", {})
        func_params: List[Dict] = self.func.get("params") or []
        self.params: List[FuncCallParam] = []

        for param in func_params:
            param_name = param.get("name")
            value = parameter_values.get(param_name, None)
            self.params.append(FuncCallParam(name=param_name, value=value, property_type=param.get("propertyType"), is_input=param.get("isInput", True)))

        self.requires_pipe = any(param.is_streamable for param in self.params)
        self.error_message = None
        self.error_stack_trace = None

    @property
    def use_parquet_transfer(self) -> bool:
        return self.options.get("isParquet", False)

    @property
    def user_api_key(self) -> Union[str, None]:
        return self.aux.get("USER_API_KEY", None)
    
    @property
    def binary_batch_size(self) -> int:
        return self.aux.get("batchSize", self.default_binary_batch_size)

    @property
    def input_params(self):
        return [x for x in self.params if x.is_input]  

    @property
    def output_params(self):
        return [x for x in self.params if not x.is_input]  

    def to_json(self):
        return {
            "id": self.id,
            "parameterValues": {p.name: p.value for p in self.params},
            "func": self.func,
            "options": self.options,
            "aux": self.aux,
            "status": self.status.value,
            "errorMessage": self.error_message,
            "errorStackTrace": self.error_stack_trace
        }
    

File: datagrok-celery-worker/src/func-call.ts
/** FuncCall model, mirroring datagrok-celery-task/func_call.py.
 *  The serialized shape is produced by the Dart FuncCall.toJson() (ddt) and consumed
 *  back by datlas' `new FuncCall.fromJson(...)` (call_queue_service.dart), so toJson()
 *  must echo `func`, `options` and `aux` untouched. */

export const Type = {
  INT: 'int',
  STRING: 'string',
  BOOL: 'bool',
  FLOAT: 'double',
  BIG_INT: 'bigint',
  DATA_FRAME: 'dataframe',
  DATE_TIME: 'datetime',
  FILE: 'file',
  BLOB: 'blob',
  GRAPHICS: 'graphics',
} as const;

export const FuncCallStatus = {
  RUNNING: 'Running',
  QUEUED: 'Queued',
  COMPLETED: 'Completed',
  ERROR: 'Error',
  CANCELED: 'Canceled',
  CANCELING: 'Canceling',
} as const;

export function isCompletedStatus(status: string): boolean {
  return status === FuncCallStatus.COMPLETED || status === FuncCallStatus.ERROR ||
    status === FuncCallStatus.CANCELED;
}

export class FuncCallParam {
  name: string;
  propertyType: string;
  isInput: boolean;
  value: any;
  originalValue: any;

  constructor(name: string, propertyType: string, isInput: boolean, value: any) {
    this.name = name;
    this.propertyType = propertyType;
    this.isInput = isInput;
    this.value = value;
    this.originalValue = value;
  }

  /** The python lib (func_call.py) treats dataframe/blob/file as streamable; the Dart
   *  server (ddt func_param.dart `isStreamable`) additionally includes graphics.
   *  We match the server so `requiresPipe` agrees with datlas' decision to open a pipe. */
  get isStreamable(): boolean {
    return this.propertyType === Type.DATA_FRAME || this.propertyType === Type.BLOB ||
      this.propertyType === Type.FILE || this.propertyType === Type.GRAPHICS;
  }
}

export class FuncCall {
  static DEFAULT_BINARY_BATCH_SIZE = 2048000;

  id: string;
  func: {[key: string]: any};
  options: {[key: string]: any};
  aux: {[key: string]: any};
  params: FuncCallParam[];
  status: string;
  errorMessage: string | null = null;
  errorStackTrace: string | null = null;

  constructor(callJson: {[key: string]: any}) {
    this.id = callJson['id'];
    this.func = callJson['func'] ?? {};
    this.options = callJson['options'] ?? {};
    this.aux = callJson['aux'] ?? {};
    this.status = FuncCallStatus.RUNNING;
    const parameterValues: {[key: string]: any} = callJson['parameterValues'] ?? {};
    const funcParams: any[] = this.func['params'] ?? [];
    this.params = funcParams.map((p) => new FuncCallParam(
      p['name'], p['propertyType'], p['isInput'] ?? true, parameterValues[p['name']] ?? null));
  }

  /** Parses a celery task message body. Accepts BOTH shapes:
   *  - the Dart hybrid `{"args": [callJson], "kwargs": {}}` produced by
   *    DockerFunc.getQueueMessageBody (docker_func.dart);
   *  - celery protocol 2 `[[callJson], {}, {...embed}]` (what a real celery
   *    client would send). */
  static fromCeleryBody(body: any): FuncCall {
    const callJson = Array.isArray(body) ? body[0]?.[0] : body?.['args']?.[0];
    if (callJson == null || typeof callJson !== 'object')
      throw new Error('Unrecognized celery task body: expected {"args": [call]} or [[call], {}, {}]');
    return new FuncCall(callJson);
  }

  get funcName(): string {
    return this.func['name'] ?? '';
  }

  get useParquetTransfer(): boolean {
    return this.options['isParquet'] === true;
  }

  get userApiKey(): string | null {
    return this.aux['USER_API_KEY'] ?? null;
  }

  get apiUrl(): string | null {
    return this.aux['DATAGROK_API_URL'] ?? null;
  }

  get binaryBatchSize(): number {
    return this.aux['batchSize'] ?? FuncCall.DEFAULT_BINARY_BATCH_SIZE;
  }

  get inputParams(): FuncCallParam[] {
    return this.params.filter((p) => p.isInput);
  }

  /** Declared outputs. */
  get outputParams(): FuncCallParam[] {
    return this.params.filter((p) => !p.isInput);
  }

  get requiresPipe(): boolean {
    return this.params.some((p) => p.isStreamable);
  }

  /** Mirrors func_call.py to_json(): inputs echo the original raw values, outputs carry
   *  the serialized value (dataframe -> {id: uuid}, blob -> param name, scalars as-is). */
  toJson(): {[key: string]: any} {
    const parameterValues: {[key: string]: any} = {};
    for (const p of this.params)
      parameterValues[p.name] = p.isInput ? p.originalValue : p.value;
    return {
      'id': this.id,
      'parameterValues': parameterValues,
      'func': this.func,
      'options': this.options,
      'aux': this.aux,
      'status': this.status,
      'errorMessage': this.errorMessage,
      'errorStackTrace': this.errorStackTrace,
    };
  }
}
